src/datetime.rs

```diff
@@ -105,6 +105,78 @@ use serde_with::{DeserializeAs, SerializeAs};
 #[derive(Eq, PartialEq, Ord, PartialOrd, Hash, Copy, Clone)]
 pub struct DateTime(i64);
 
+/// Struct providing a builder to construct a `bson::DateTime` object from a given year, month, and
+/// date, and optionally the hour, minute, second and millisecond, which default to 0 if not
+/// explicitly set.
+///
+/// ```
+/// use bson::{DateTime, datetime::DateTimeBuilder};
+///
+/// let dt = DateTimeBuilder::new(1998, 2, 12).minute(1).millisecond(23).builder();
+/// let expected = DateTime::parse_rfc3339_str("1998-02-12T00:01:00.023Z").unwrap();
+/// assert_eq!(dt.unwrap(), expected);
+/// ```
+pub struct DateTimeBuilder {
+    year: i32,
+    month: u8,
+    day: u8,
+    hour: Option<u8>,
+    minute: Option<u8>,
+    second: Option<u8>,
+    millisecond: Option<u16>,
+}
+
+impl DateTimeBuilder {
+    pub fn new(year: i32, month: u8, day: u8) -> Self {
+        Self {
+            year,
+            month,
+            day,
+            hour: None,
+            minute: None,
+            second: None,
+            millisecond: None,
+        }
+    }
+
+    pub fn hour(&mut self, hour: u8) -> &mut Self {
+        self.hour = Some(hour);
+        self
+    }
+
+    pub fn minute(&mut self, minute: u8) -> &mut Self {
+        self.minute = Some(minute);
+        self
+    }
+
+    pub fn second(&mut self, second: u8) -> &mut Self {
+        self.second = Some(second);
+        self
+    }
+
+    pub fn millisecond(&mut self, millisecond: u16) -> &mut Self {
+        self.millisecond = Some(millisecond);
+        self
+    }
+
+    pub fn builder(&mut self) -> Result<DateTime> {
```

Comment:
nit: can we call this method `build`? that's the naming convention for the builders we have in the driver